web/src/engine/websites/KLMangaSh.ts

```diff
@@ -0,0 +1,90 @@
+import icon from './KLMangash.webp';
+import { type Chapter, DecoratableMangaScraper, Page } from '../providers/MangaPlugin';
+import * as Common from './decorators/Common';
+import { FetchHTML, FetchJSON, FetchWindowScript } from '../platform/FetchProvider';
+
+type ZingParams = {
+    nonce: string,
+    ajax_url: string
+}
+
+type PageResult = {
+    mes: string,
+    going: number,
+    img_index: number
+}
+
+function CleanTitle(title: string): string {
+    return title.replace(/\(Raw.*Free\)/i, '').trim();
+}
+
+function MangaLabelExtractor(element: HTMLElement) : string{
+    return CleanTitle(element.textContent);
+}
+function MangaExtractor(anchor: HTMLAnchorElement) {
+    return {
+        id: anchor.pathname,
+        title: CleanTitle(anchor.text)
+    };
+}
+
+function ChapterExtractor(anchor: HTMLAnchorElement) {
+    return {
+        id: anchor.pathname,
+        title: CleanTitle(anchor.querySelector<HTMLSpanElement>('span').textContent)
+    };
+}
+
+@Common.MangaCSS(/^{origin}\/manga-raw\/[^/]+\/$/, 'div.container div.z-single-mg h1.name', MangaLabelExtractor)
+@Common.MangasMultiPageCSS('/page/{page}/', 'div.grid-of-mangas h2.name a', 1, 1, 0, MangaExtractor)
+@Common.ChaptersSinglePageCSS('div.chapter-box a', ChapterExtractor)
+@Common.ImageAjax(true)
+export default class extends DecoratableMangaScraper {
+    private zingParams: ZingParams;
+
+    public constructor() {
+        super('klmangash', 'KLManga(.sh)', 'https://klmanga.sh', Tags.Language.Japanese, Tags.Source.Aggregator);
+    }
+
+    public override get Icon() {
+        return icon;
+    }
+
+    public override async Initialize(): Promise<void> {
+        this.zingParams = await FetchWindowScript(new Request(this.URI), 'new Promise (resolve => resolve({ nonce: zing.nonce_a, ajax_url : zing.ajax_url }));', 500);
+    }
+
+    public override async FetchPages(chapter: Chapter): Promise<Page[]> {
+        const pages: Page[] = [];
+        const doc = await FetchHTML(new Request(new URL(chapter.Identifier, this.URI)));
+        const p = doc.documentElement.innerHTML.match(/\sp:\s*(\d+),/).at(1);
+        const chapter_id = doc.documentElement.innerHTML.match(/chapter_id\s*:\s*['"]([^'"]+)/).at(1);
+
+        const params = new URLSearchParams({
+            nonce_a: this.zingParams.nonce,
+            action: 'z_do_ajax',
+            _action: 'decode_images_100',
+            p,
+            img_index: '0',
+            chapter_id,
+            content: ''
+        });
+
+        for (let run = true; run;) {
+            const request = new Request(this.zingParams.ajax_url, {
+                method: 'POST',
+                body: params,
+                headers: {
+                    'Content-type': 'application/x-www-form-urlencoded; charset=UTF8',
+                    'X-Requested-With': 'XMLHttpRequest',
+                }
+            });
+            let { img_index, mes, going } = await FetchJSON<PageResult>(request);
```

Comment:
The `FetchPages` method must avoid making too many requests.
Either parallelize the requests, or better move the logic to `FetchImage`.

Reply:
Ikr, but we DONT KNOW the real number of pages so both of your solutions are impossible.
The real number of pages is known once the website api return going = 0 ie. after the loop.

Crappy website coding.

Reply:
I've checked rawlazy before submitting that PR, since i saw the similarities. Rawlazy is simply using lazy loading, with one img element per page. Here we have nothing afaik.

Reply:
I mean there is a dirty way. Make the request once, you get first two pages.
Then we have the page folder.

`hoster.com/path/number.extension.`

Since the pages urls follow a pattern we could send HEAD requests while incrementing number.

Reply:
May find the last page by using divide and conquer algorithm, within 6 requests it would be possible to find the last API call which provides a non empty content (assuming there are < 128 pages).

You may leave as is for now, may take a look into it later